### PR TITLE
Fix http1 chunked upload

### PR DESCRIPTION
**Pull request by HeinrichHartmann**

### CODE REVIEW

File: src/mtev_http1.c [hunk 1]
```diff
@@ -835,12 +837,15 @@ mtev_http1_complete_request(mtev_http1_session_ctx *ctx, int mask, struct timeva
...
+    mtevL(http_debug, "[fd=%d] mtev_http -> read() = %d\n", CTXFD(ctx), len);
     if(len > 0)
-      mtevL(http_io, " mtev_http:read(%d) => %d [\n%.*s\n]\n", CTXFD(ctx), len, len, in->buff + in->start + in->size);
+      mtevL(http_io, "[fd=%d] mtev_http:read() => %d [\n%.*s\n]\n", CTXFD(ctx), len, len, in->buff + in->start + in->size);
     else
-      mtevL(http_io, " mtev_http:read(%d) => %d\n", CTXFD(ctx), len);
-    if(len == -1 && errno == EAGAIN) return mask;
+      mtevL(http_io, "[fd=%d] mtev_http:read() => %d\n", CTXFD(ctx), len);
+    if(len == -1 && errno == EAGAIN) {
+      sleep(1);
```
Comment (postwait) [inline src/mtev_http1.c]: no sleep

File: src/mtev_http1.c [hunk 2]
```diff
@@ -996,6 +1001,24 @@ mtev_http1_session_req_consume_read(mtev_http1_session_ctx *ctx,
...
+      if(head->allocd - head->start < MAX_CHUNK_HEADER_SIZE) {
+        mtevL(http_debug, "[fd=%d] Too little space in head node (%d). Reallocating.\n",
+              CTXFD(ctx), (int) (head->allocd - head->start));
+        struct bchain *new_head = ALLOC_BCHAIN(DEFAULT_BCHAINSIZE);
+        memcpy(new_head->buff, head->buff + head->start, head->size);
+        new_head->size = head->size;
+        new_head->start = 0;
+        new_head->compression = head->compression;
+        new_head->next = head->next;
+        /* It's safe to set {first,last}_input here, see (*@L1087) below. */
```
Comment (postwait) [inline src/mtev_http1.c]: don't use a line number say "see the 'expand req.*_input list' comment below"

File: src/mtev_http1.c [hunk 3]
```diff
@@ -996,6 +1001,24 @@ mtev_http1_session_req_consume_read(mtev_http1_session_ctx *ctx,
       }
     }
     else {
+      /* Check if we have enough in this node read the chunk header "$chunk_size\r\n" */
```
Comment (esproul) [inline src/mtev_http1.c]: I don't understand this comment. Perhaps it should read "enough space in this node to read"?

File: src/mtev_http1.c [hunk 4]
```diff
@@ -1052,6 +1082,17 @@ mtev_http1_session_req_consume_read(mtev_http1_session_ctx *ctx,
           ALLOC_BCHAIN(DEFAULT_BCHAINSIZE);
     }
     else if (tail->start + tail->size >= tail->allocd) {
+      /* expand req.*_input list
+       *
+       * (*) At least in the chunked case, we never get here.
+       * This code is executed only if the function is called with a full buffer before reading from
+       * the socket (below). However,
+       * - If we were holding a complete chunk in the buffer, we would have jumped this section (`goto` ~ L1038).
+       * - If we were holding an incomplete chunk, we expand the buffer to make it fit (`ALLOC_BCHAIN` ~ L1040).
```
Comment (postwait) [inline src/mtev_http1.c]: Again, these line number references are fragile.  Use a more resilient reference method... C[1] C[2] is possible.

File: src/mtev_http1.c (hunk 1 above)
Comment (HeinrichHartmann) [reply src/mtev_http1.c]: uh. Indeed!

File: src/mtev_http1.c (hunk 3 above)
Comment (HeinrichHartmann) [reply src/mtev_http1.c]: fixed

File: src/mtev_http1.c (hunk 4 above)
Comment (HeinrichHartmann) [reply src/mtev_http1.c]: fixed

File: src/mtev_http1.c [hunk 5]
```diff
@@ -996,6 +1000,24 @@ mtev_http1_session_req_consume_read(mtev_http1_session_ctx *ctx,
       }
     }
     else {
+      /* Check if we have enough space in this bchain node for the chunk header "$chunk_size\r\n" */
+      if(head->allocd - head->start < MAX_CHUNK_HEADER_SIZE) {
+        mtevL(http_debug, "[fd=%d] Too little space in head node (%d). Reallocating.\n",
+              CTXFD(ctx), (int) (head->allocd - head->start));
+        struct bchain *new_head = ALLOC_BCHAIN(DEFAULT_BCHAINSIZE);
```
Comment (postwait) [inline src/mtev_http1.c]: I wonder if allocd is large enough, could we just memmove the start back to 0 in the same bchain without allocating a new one and freeing the old one.  The fact that we free the old one leads me to "yes".

Comment (HeinrichHartmann) [reply src/mtev_http1.c]: I think that's true. Let me fix this.

Comment (HeinrichHartmann) [reply src/mtev_http1.c]: Updated.

File: src/mtev_http1.c [hunk 6]
```diff
@@ -996,6 +1000,15 @@ mtev_http1_session_req_consume_read(mtev_http1_session_ctx *ctx,
       }
     }
     else {
+      /* Check if we have enough space in this bchain node for the chunk header "$chunk_size\r\n" */
+      if(head->allocd - head->start < MAX_CHUNK_HEADER_SIZE) {
+        mtevL(http_debug, "[fd=%d] Too little space in head node (%d). Resetting bchain node.\n",
+              CTXFD(ctx), (int) (head->allocd - head->start));
+        memmove(head->buff, head->buff + head->start, head->size);
+        head->size = head->size;
```
Comment (postwait) [inline src/mtev_http1.c]: noop

File: src/mtev_http1.c [hunk 7]
```diff
@@ -996,6 +1000,15 @@ mtev_http1_session_req_consume_read(mtev_http1_session_ctx *ctx,
       }
     }
     else {
+      /* Check if we have enough space in this bchain node for the chunk header "$chunk_size\r\n" */
+      if(head->allocd - head->start < MAX_CHUNK_HEADER_SIZE) {
```
Comment (postwait) [inline src/mtev_http1.c]: I think you still need the old code if this predicate fails.  And this is <= no?

File: src/mtev_http1.c (hunk 6 above)
Comment (HeinrichHartmann) [reply src/mtev_http1.c]: indeed...

File: src/mtev_http1.c (hunk 7 above)
Comment (HeinrichHartmann) [reply src/mtev_http1.c]: Regarding `<=`
- If we have equality then, the header should fit and we are good.
- In any case, MAX_CHUNK_HEADER_SIZE == 20, is just a guess, so `<` vs. `<=` is not an important edge case to consider.

If this condition fails, the current bchain node is large enough for our needs. The subsequent reads will be able to get the chunk header field into the buffer.
Once we have this, the section `if(str_in_f) { ... }` will prepare for the chunk payload.

**Comment by HeinrichHartmann** [comment]
Expanding on the last comment. The whole loop could be redone, using a single large buffer that holds the latest chunk. Once we have completely read it, we move it to the user_data linked list (as we do now).
Looks like the {first,last}_input list is not doing much.

The reason I did not go for this refactor is, that I don't understand:
- (a) the non-chunked case
- (b) the reading of header fields, etc. in other http_request_* functions.

**Comment by HeinrichHartmann** [comment]
Tests pass locally.

**Pull request merged**